app/src/main/cpp/skyline/loader/nro.cpp

```diff
@@ -1,38 +1,65 @@
 // SPDX-License-Identifier: MPL-2.0
 // Copyright © 2020 Skyline Team and Contributors (https://github.com/skyline-emu/)
 
-#include <vector>
+#include <nce.h>
+#include <os.h>
 #include <kernel/memory.h>
+#include <vfs/nacp.h>
+#include <vfs/region_backing.h>
 #include "nro.h"
 
 namespace skyline::loader {
-    NroLoader::NroLoader(int fd) : Loader(fd) {
-        ReadOffset((u32 *) &header, 0x0, sizeof(NroHeader));
+    NroLoader::NroLoader(const std::shared_ptr<vfs::Backing> &backing) : Loader(backing) {
+        backing->Read(&header);
 
         if (header.magic != util::MakeMagic<u32>("NRO0"))
             throw exception("Invalid NRO magic! 0x{0:X}", header.magic);
+
+        // The homebrew asset section is appended to the end of an NRO file
+        if (backing->size > header.size) {
+            backing->Read(&assetHeader, header.size);
+
+            if (assetHeader.magic != util::MakeMagic<u32>("ASET"))
+                throw exception("Invalid ASET magic! 0x{0:X}", assetHeader.magic);
+
+            NroAssetSection &nacpHeader = assetHeader.nacp;
+            nacp = std::make_shared<vfs::NACP>(std::make_shared<vfs::RegionBacking>(backing, header.size + nacpHeader.offset, nacpHeader.size));
+        }
     }
 
-    void NroLoader::LoadProcessData(const std::shared_ptr<kernel::type::KProcess> process, const DeviceState &state) {
-        std::vector<u8> text(header.text.size);
-        std::vector<u8> rodata(header.ro.size);
-        std::vector<u8> data(header.data.size);
+    std::vector<u8> NroLoader::GetIcon() {
+        NroAssetSection &segmentHeader = assetHeader.icon;
+        std::vector<u8> buffer(segmentHeader.size);
+
+        backing->Read(buffer.data(), header.size + segmentHeader.offset, segmentHeader.size);
+        return buffer;
```

Comment:
In certain cases the `ASET` section may be missing or the offset/size specified in the `icon` header can be 0. Check [NroLoader.kt](https://github.com/skyline-emu/skyline/blob/4d787c904e12e671c1ff115938b548e79b492130/app/src/main/java/emu/skyline/loader/NroLoader.kt#L21) for reference, you should return a 0-sized vector here in case the icon doesn't exist.

Reply:
oh right, forgot about that

Reply:
hmm actually this is fine, since assetHeader is zeroed by default and read will accept zero byte reads

Reply:
If the `ASET` header isn't there then that part could be filled with random data rather than just 0s.

Reply:
Nope, if asset header is missing then we don't change the assetHeader from the default, which is being filled with 0s.

Reply:
:shrek: